src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs

```diff
@@ -36,6 +40,96 @@ public static IServiceCollection AddMvc([NotNull] this IServiceCollection servic
             services.Configure(setupAction);
         }
 
+        /// <summary>
+        /// Register the specified <paramref name="controllerTypes"/> as services and as a source for controller
+        /// discovery.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="controllerTypes">A sequence of controller <see cref="Type"/>s to register in the <paramref name="services"/>
+        /// and used for controller discovery.</param>
+        /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        public static IServiceCollection WithControllersAsServices(
+           [NotNull] this IServiceCollection services,
+           [NotNull] IEnumerable<Type> controllerTypes)
+        {
+            return WithControllersAsServices(services, controllerTypes, configuration: null);
+        }
+
+        /// <summary>
+        /// Register the specified <paramref name="controllerTypes"/> as services and as a source for controller
+        /// discovery.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="controllerTypes">A sequence of controller <see cref="Type"/>s to register 
+        /// in the <paramref name="services"/> and used for controller discovery.</param>
+        /// <param name="configuration">The application's <see cref="IConfiguration"/>.</param>
+        /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        public static IServiceCollection WithControllersAsServices(
+            [NotNull] this IServiceCollection services,
+            [NotNull] IEnumerable<Type> controllerTypes,
+            IConfiguration configuration)
+        {
+            var controllerTypeInfos = new List<TypeInfo>();
+            foreach (var type in controllerTypes)
+            {
+                services.AddTransient(type);
+                controllerTypeInfos.Add(type.GetTypeInfo());
+            }
+
+            var describer = new ServiceDescriber(configuration);
+            services.Replace(describer.Transient<IControllerActivator, ServiceBasedControllerActivator>());
+            var controllerTypeProvider = new StaticControllerTypeProvider(controllerTypeInfos);
+            services.Replace(describer.Instance<IControllerTypeProvider>(controllerTypeProvider));
```

Comment:
I'll run it by Louis to see if this wouldn't require the config replacement behavior since you opted to call this method rather than the scenario with AddMvc where you get services wholesale.